include/circt/Support/PrettyPrinterHelpers.h

```diff
@@ -166,6 +166,46 @@ class TokenStringSaver : public PrettyPrinter::Listener {
   void clear() override;
 };
 
+template <typename Callable, typename Data>
+class CallbackSaver {
+
+  /// List of all the unique data associated with each callback token.
+  SmallVector<Data *> list;
+  /// The storage for the callback, as a function object.
+  std::function<void(unsigned)> callbackObj;
+
+  /// Note: Callable class must implement a callable with signature::
+  /// void (Data*)
+public:
+  CallbackSaver() = default;
```

Comment:
Maybe this shouldn't have a default constructor (and `init()` made private)?  I say this because as-is these methods make it easy to use this badly-- either with no callable, or replacing an existing callable.